README.md
ABOUT
==========
This is version 2.4.x of the "Attendance" module (attforblock).

2.4.x release is compatible only with Moodle 2.x.

The "Attendance" module is developed by
    Dmitry Pupinin, Novosibirsk, Russia,
    Artem Andreev, Taganrog, Russia.

This block may be distributed under the terms of the General Public License
(see http://www.gnu.org/licenses/gpl.txt for details)

PURPOSE
==========
The attendance module and block are designed to allow instructors of a course keep an attendance log of the students in their courses. The instructor will setup the frequency of his classes (# of days per week & length of course) and the attendance is ready for use. To take attendance, the instructor clicks on the "Update Attendance" button and is presented with a list of all the students in that course, along with 4 options: Present, Absent, Late & Excused, with a Remarks textbox. Instructors can download the attendance for their course in Excel format or text format.
Only the instructor can update the attendance data. However, a student gets to see his attendance record.

INSTALLATION
==========
The attendance module follows standard installation procedure.

1. Create folder <path to your moodle dir>/mod/attforblock.
2. Extract files from folder inside archive (andreev-artem-moodle_mod_attforblock-xxxxxxx in your case) to created folder.
3. Visit page Home ► Site administration ► Notifications to complete installation.

CHANGES
2011-07-16 andreev.artem

        2.4 beta contains such changes:
        * compatible only with Moodle 2.0
        * fully internals rewritten (OOP, renderers, etc)
        * support all 2.3 features: session types (common and group),
        interface improvements (sessions filtering by different criteria’s),
        copy attendance data from same-day sessions,
        fast session switching for same-day sessions,
        more than one attendance in the same course,
        list and grid display modes for take page
        * enhancement: taken into account information
        about students enrolment (start/end date) on different pages
        * enhancement: HTML-editor for session description
        with ability to add links, files, formatting, etc

2011-01-18  lrlopez

        * attendances.php: styles.php:

        Implemented copy attendance data from same-day sessions.
        Fixed fast session dropdown list end time inaccuracy.
        Enhanced session dropdown list.

        * lang/: en_utf8/attforblock.php, es_utf8/attforblock.php:

        Added new strings for the copy dropdown menus.

2011-01-18  lrlopez

        * attendances.php: styles.php:

        Implemented fast session switching for same-day sessions.
        Fixed some XHTML compliance tags (more coming)

        * lang/: en_utf8/attforblock.php, es_utf8/attforblock.php:

        Added new strings for the dropdown menus.

2011-01-16  lrlopez

        * export.php: export_form.php:
        * lang/: en_utf8/attforblock.php: es_utf8/attforblock.php:
        Added "Start of period" into export tab and two new strings for
        the new fields on the form.

2011-01-15  lrlopez

        * db/install.xml:

        Bugs fixed.

        * attendances.php: locallib.php: manage.php:

        Fixed some minor bugs and code readibility issues. Now, in grid view
        mode, statuses are always next to their radio buttons.

2010-12-30 - 2011-01-04  lrlopez

        * lang/: es_utf8/attforblock.php:

        Updated and fixed Spanish translation (Enrique Robredo)

        * styles.php:

        Fixed yui-calendar CSS modifying global CSS behaviour.

        * manage.php:

        Added an option for choosing between show session duration or show end
        time in session list.

        * db/: install.xml: upgrade.php: version.php:
        * attsettings.php: lib.php: locallib.php: manage.php: report.php:

        Added support to allow more than one attendance in the same course.

        * sessions.php:

        Fixed DST bug (adding multiple sessions could generate invalid start
        times if any DST change occur in-between)

        * attendances.php:
        Implemented display modes (list and grid)

        * lang/: en_utf8/attforblock.php, es_utf8/attforblock.php:
        Added new strings for the new display modes

2010-07-10 - 2010-09-17  andreev.artem

    Implemented ability to add two types of sessions: common and group

    Some interface improvements (sessions filtering by different criteries)
    for manage.php, report.php, view.php

    Don't display grade and percent when module grade option is "No grade"

    For redirecting from the block is taken into account the situation
    when the current time are a few sessions

2009-04-13 Monday 01:50  dlnsk

	* lang/: en_utf8/attforblock.php, ru_utf8/attforblock.php: 
	
	two forgotten strings

2009-04-13 Monday 01:48  dlnsk

	* locallib.php: 
	
	array_intersect_key which require PHP 5.1.0 replaced to local
	implementation

2009-04-11 Saturday 12:21  dlnsk

	* README.txt: 
	
	installation description fixed

2009-04-08 Wednesday 01:23  dlnsk

	* lang/ja_utf8/help/attforblock/hiddensessions.html: 
	
	file hiddensessions.html was initially added on branch
	MOODLE_19_STABLE.

2009-04-08 Wednesday 01:23  dlnsk

	* lang/ja_utf8/: attforblock.php,
	help/attforblock/createmultiplesessions.html,
	help/attforblock/createonesession.html,
	help/attforblock/defaultstatuses.html,
	help/attforblock/display.html, help/attforblock/export.html,
	help/attforblock/myvariables.html, help/attforblock/report.html,
	help/attforblock/sessions.html, help/attforblock/studentview.html,
	help/attforblock/hiddensessions.html: 
	
	Japanese language updated Thanks to Mitsuhiro Yoshida

2009-03-12 Thursday 03:21  dlnsk

	* lib.php: 
	
	bug fixed: notofication rise when module adding into course.
	Updating gradebook is a reason. Now Attendance show up in gradebook
	only after taking attendance.

2009-03-12 Thursday 03:17  dlnsk

	* export_form.php: 
	
	bug fixed: 'no groups' cause notification

2009-03-10 Tuesday 20:19  dlnsk

	* lang/cs_utf8/attforblock.php: 
	
	Czech Language. Thanks to Lukas Mizoch

2009-03-10 Tuesday 20:19  dlnsk

	* lang/cs_utf8/attforblock.php: 
	
	file attforblock.php was initially added on branch
	MOODLE_19_STABLE.

2009-03-01 Sunday 04:20  dlnsk

	* lang/en_utf8/help/attforblock/hiddensessions.html: 
	
	file hiddensessions.html was initially added on branch
	MOODLE_19_STABLE.

2009-03-01 Sunday 04:20  dlnsk

	* RELEASES.txt, backuplib.php, manage.php, restorelib.php,
	version.php, lang/en_utf8/attforblock.php,
	lang/en_utf8/help/attforblock/hiddensessions.html,
	lang/ru_utf8/attforblock.php: 
	
	Format of backup changed Hidden sessions displayed Small fixes

2009-03-01 Sunday 01:49  dlnsk

	* report.php: 
	
	bug fix: statuses on Report page

2009-02-27 Friday 21:08  dlnsk

	* locallib.php: 
	
	bug fixed: devision by zero

2009-02-25 Wednesday 20:07  dlnsk

	* lang/es_utf8/: attforblock.php,
	help/attforblock/createmultiplesessions.html,
	help/attforblock/createonesession.html,
	help/attforblock/display.html, help/attforblock/export.html,
	help/attforblock/myvariables.html, help/attforblock/mods.html,
	help/attforblock/report.html, help/attforblock/sessions.html,
	help/attforblock/studentview.html,
	help/attforblock/updateattendance.html: 
	
	Updated Thanks to Enrique Robredo.

2009-02-25 Wednesday 19:30  dlnsk

	* lang/es_utf8/help/attforblock/updateattendance.html: 
	

_view.php
<?PHP  // $Id: view.php,v 1.3.2.2 2009/02/23 19:22:41 dlnsk Exp $

/// This page prints a particular instance of attforblock
/// (Replace attforblock with the name of your module)

    require_once("../../config.php");
    require_once('locallib.php');

    $id   = optional_param('id', -1, PARAM_INT);   // Course Module ID, or
//    $a    = optional_param('a', -1, PARAM_INT);    // attforblock ID
	$studentid			= optional_param('student', 0, PARAM_INT);
	$printing			= optional_param('printing', 0, PARAM_INT);
    $mode 				= optional_param('mode', 'thiscourse', PARAM_ALPHA);
    $view       = optional_param('view', NULL, PARAM_ALPHA);        // which page to show
	$current	= optional_param('current', 0, PARAM_INT);
	
    if ($id) {
        if (! $cm = $DB->get_record("course_modules", array("id"=> $id))) {
            error("Course Module ID was incorrect");
        }
    
        if (! $course = $DB->get_record("course", array("id"=> $cm->course))) {
            error("Course is misconfigured");
        }
    
        if (! $attforblock = $DB->get_record("attforblock", array("id"=> $cm->instance))) {
            error("Course module is incorrect");
        }

    } else {
    	error("Module id is incorrect.");
//        if (! $attforblock = get_record("attforblock", "id", $a)) {
//            error("Course module is incorrect");
//        }
//        if (! $course = get_record("course", "id", $attforblock->course)) {
//            error("Course is misconfigured");
//        }
//        if (! $cm = get_coursemodule_from_instance("attforblock", $attforblock->id, $course->id)) {
//            error("Course Module ID was incorrect");
//        }
    }

    require_login($course->id);

    if (!$context = get_context_instance(CONTEXT_MODULE, $cm->id)) {
        print_error('badcontext');
    }
    
    if (!$studentid && (has_capability('mod/attforblock:manageattendances', $context) ||
                        has_capability('mod/attforblock:takeattendances', $context) ||
                        has_capability('mod/attforblock:changeattendances', $context))) {
        redirect("manage.php?id=$cm->id");
    }
    if (!$studentid && has_capability('mod/attforblock:viewreports', $context)) {
        redirect("report.php?id=$cm->id");
    }
    
    if (! $user = $DB->get_record('user', array('id'=> $USER->id) )) {
        error("No such user in this course");
    }
	
    if ($view)
        set_current_view($course->id, $_GET['view']);
    else
	    $view = get_current_view($course->id, 'months');

    require_capability('mod/attforblock:view', $context);
    
	$student = false;
    if ($studentid) {
    	if ($studentid == $USER->id or has_capability('mod/attforblock:viewreports', $context)) {
		    if (!$student = $DB->get_record('user', array('id'=> $studentid) )) {
		        error("No such user in this course");
		    }
    	}
	}
	
//    if (empty($student) && has_capability('mod/attforblock:manageattendances', $context)) {
//        redirect("manage.php?id=$cm->id");
//    }
    
	if ($student) {
		$user = $student;
	}
	if ($printing) {
		if ($mode === 'thiscourse') {
    		print_header('', $course->fullname.' - '.get_string('attendancereport','attforblock'));
			print_user_attendaces($user, $cm, $attforblock, $course, 'printing');
		} else {
    		print_header('', get_string('attendancereport','attforblock'));
			print_user_attendaces($user, $cm, $attforblock, 0, 'printing');
		}
    	exit();
    }

/// Print headers
	    $navlinks[] = array('name' => $attforblock->name, 'link' => "view.php?id=$id", 'type' => 'activityinstance');
	    $navlinks[] = array('name' => get_string('attendancereport', 'attforblock'), 'link' => null, 'type' => 'title');
	    $navigation = build_navigation($navlinks);
	    print_header("$course->shortname: ".$attforblock->name.' - ' .get_string('export', 'quiz'), $course->fullname,
	                 $navigation, "", "", true, "&nbsp;", navmenu($course));

	//add info to log
	add_to_log($course->id, 'attendance', 'student view', "mod/attforblock/view.php?course=$course->id&amp;student=$USER->id", $USER->lastname.' '.$USER->firstname);
//	print_heading(get_string('attendanceforthecourse','attforblock').' :: ' .$course->fullname);

	/// Prints out tabs
    $currenttab = $mode;
    $studstr = $student ? '&amp;student='.$student->id : '';
    $toprow = array();
    $toprow[] = new tabobject('thiscourse', "view.php?id=$id&amp;mode=thiscourse{$studstr}",
                get_string('thiscourse','attforblock'));
    $toprow[] = new tabobject('allcourses', "view.php?id=$id&amp;mode=allcourses{$studstr}",
                   get_string('allcourses','attforblock'));
    print_tabs(array($toprow), $currenttab);
    
	if ($mode === 'thiscourse') {
		print_user_attendaces($user, $cm, $attforblock, $course);
	} else {
    	print_user_attendaces($user, $cm, $attforblock);
	}

	print_footer($course);

?>